Fix temperature error code despite success status code

lisabiermann opened the pull request:
## Description

This PR fixes #173.

## Type of Change

<!-- Please select only a single option with an "x" or "X". -->

 - [ ] Break (major)
 - [ ] Feature (minor)
 - [x] Bug fix (patch)
 - [ ] No source changes (no version change)

Commit:
Apply black

lisabiermann committed:
add mass over temperature ratio check output

lisabiermann committed:
unify conditions for CalcTempAtFalseVacFraction

vollous committed:
Merge branch 'master' into develop

vollous commented:
Would you mind creating a unit test with Christoph's point so that this never happens again? In case we change the logic in the future.

vollous committed:
Merge branch 'master' into develop

vollous committed:
Move negative bosons spline into the .cpp files (#182)

* Fix

* Fix windows

lisabiermann committed:
Merge branch 'develop' into fix_Temp_from_VacFraction

lisabiermann committed:
Add unit test to check new logic

lisabiermann committed:
Merge branch 'fix_Temp_from_VacFraction' of github.com:phbasler/BSMPT into fix_Temp_from_VacFraction

vollous committed:
Mathematica model implementation (#171)

* Mathematica implementation

* Small improvements

* Automate model implementation (src still not implemented)

* N2HDM full mathematica implementation

* nQuark nGauge nLepton different from SM support. Still has to be tested.

* Update tools/ModelGeneration/Mathematica/N2HDM.m

Co-authored-by: Lisa Biermann <51697431+lisabiermann@users.noreply.github.com>

* Update tools/ModelGeneration/Mathematica/CreateModel.m

Co-authored-by: Lisa Biermann <51697431+lisabiermann@users.noreply.github.com>

* Remove "21" to nRows of EqMatrix

* Import Helper functions at the end

* Remove some errors

* Move helper function to the beginning

* Moved "InputParameters"

* Fix test not passing. All dependent vars need to be defined as function of the input parameters.

* More appropriate name

* Example file is created right away with some values set by the user

* Removing warning

* Add reference for implementation and add R2HDM

---------

Co-authored-by: Lisa Biermann <51697431+lisabiermann@users.noreply.github.com>
Co-authored-by: Lisa Biermann <li.biermann@web.de>

lisabiermann committed:
Rename tmp in a meaningful way

lisabiermann committed:
Merge branch 'develop' into highT_validity_check

lisabiermann committed:
Add max ratio output and unit test

lisabiermann committed:
Merge pull request #175 from phbasler/highT_validity_check

Add mass over temperature ratio check output

vollous committed:
Update develop branch (#190)

* Dont try to autocommit on dependabot or fork jobs (#187)

* Dont use secret in dependabot job

* add brackets

* trigger change

* Automatically applied linter

* print env

* add brackets

* Automatically applied linter

* remove debug line

---------

Co-authored-by: GitHub Actions Bot <>

* Next attempt at fixing linter for non local (#188)

* Next attempt at fixing linter for non local

* Update linter.yml

* Update .github/workflows/linter.yml

* Update .github/workflows/linter.yml

* Bump JamesIves/github-pages-deploy-action from 4.6.9 to 4.7.2 (#186)

* Bump JamesIves/github-pages-deploy-action from 4.6.9 to 4.7.2

Bumps [JamesIves/github-pages-deploy-action](https://github.com/jamesives/github-pages-deploy-action) from 4.6.9 to 4.7.2.
- [Release notes](https://github.com/jamesives/github-pages-deploy-action/releases)
- [Commits](https://github.com/jamesives/github-pages-deploy-action/compare/v4.6.9...v4.7.2)

---
updated-dependencies:
- dependency-name: JamesIves/github-pages-deploy-action
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

* Next attempt at fixing linter for non local

* Update linter.yml

* Update linter.yml

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Philipp Basler <28863303+phbasler@users.noreply.github.com>

* Remove test for ubuntu 20.04 (#189)

---------

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: Philipp Basler <28863303+phbasler@users.noreply.github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

vollous committed:
Merge branch 'develop' into fix_Temp_from_VacFraction

vollous committed:
Attempt at decreasing computation time on test "Check percolation temperature for point of issue #173"

vollous committed:
Merge branch 'master' into develop

vollous reviewed (dismissed):
Looks good. I added some changes to the test to make it quicker (we were getting timeouts).

vollous committed:
155 fix dependency loop between models and minimizer (#183)

* Objective

* Automated commit of cmake-format changes.

* Fix

* Automated commit of cmake-format changes.

* modelPointer is no longer passed as a pointer

* Check CheckImplementation

* Automatically applied linter

* Missing lib

* add a check that the unit test does check the result text for the pass substring

---------

Co-authored-by: vollous <vollous@users.noreply.github.com>
Co-authored-by: Philipp Basler <28863303+phbasler@users.noreply.github.com>
Co-authored-by: GitHub Actions Bot <>

vollous committed:
Merge branch 'master' into develop

Commit:
Automatically applied linter

vollous committed:
Merge branch 'master' into develop

phbasler committed:
Use repository option to fix linter for forks (#203)

phbasler committed:
Merge branch 'develop' into fix_Temp_from_VacFraction

phbasler commented on src/bounce_solution/bounce_solution.cpp:
```diff
@@ -754,14 +754,11 @@ double BounceSolution::CalcTempAtFalseVacFraction(const double &false_vac_frac)
-                   MarginOfCalcTempAtFalseVacFractionBeforeFailure)))
+    while (not(std::abs(T_up / T_down - 1) <
+                   RelativeTemperatureInCalcTempAtFalseVacFraction and
+               almost_the_same(int_at_false_vac_frac,
```
In the unit test the almost_same part of the while condition is never hit.
Is this just a because the choice of test point or is this in general only a fallback check?

cborschensky commented:
There were two things regarding the changes in the PR that I noticed:

### Comment 1
The logic of [lines 757-764](https://github.com/phbasler/BSMPT/blob/0aea6b73e6dd2165996b92ad873e54917b8c419c/src/bounce_solution/bounce_solution.cpp#L757) changed. Is this intended? While before, it was of the type `while (A > B and not C)` with
```
A = std::abs(T_up / T_down - 1)
B = RelativeTemperatureInCalcTempAtFalseVacFraction * MarginOfCalcTempAtFalseVacFractionBeforeFailure
C = almost_the_same(int_at_false_vac_frac, IatT, RelativeErrorInCalcTempAtFalseVacFraction * MarginOfCalcTempAtFalseVacFractionBeforeFailure)
```
this PR changes it to `while not (A < B' and C')` with
```
B' = RelativeTemperatureInCalcTempAtFalseVacFraction
C' = almost_the_same(int_at_false_vac_frac, IatT, RelativeErrorInCalcTempAtFalseVacFraction)
```
which is something like `while (A > B' or not C')` when distributing the `not`, so, ignoring the differences between `B` and `B'` as well as `C` and `C'`, there is now an `or` instead of an `and` operator, requiring now both conditions to be fulfilled before the loop exits. I guess this is also related to @phbasler's https://github.com/phbasler/BSMPT/pull/174#discussion_r2013079692.

This change in logic can actually cause an infinite loop, because the integral `IatT` in [line 768](https://github.com/phbasler/BSMPT/blob/0aea6b73e6dd2165996b92ad873e54917b8c419c/src/bounce_solution/bounce_solution.cpp#L768) is not always calculated precisely enough during the bisection, so in such a case, the `almost_the_same(...)` condition is never true! For the example [input_infinite_loop_2.txt](https://github.com/user-attachments/files/19507569/input_infinite_loop_2.txt) with the command line
```
bin/CalcTemps --model=r2hdm --input=input_infinite_loop_2.txt --output=test.tsv --firstline=2 --lastline=2 --logginglevel::bouncedetailed=true
```
I found such a case where the integral `IatT` seems to be very steep, and its value always switches between `0` and a very large value, but never reaches or gets close to the desired value `int_at_false_vac_frac`, and so the infinite loop occurs (the point seems to be a bit unstable, however, as playing around with the settings of the minimizers makes BSMPT exit quickly sometimes, saying the point is not NLO stable).

### Comment 2
The condition of the `while` loop in [lines 757-764](https://github.com/phbasler/BSMPT/blob/0aea6b73e6dd2165996b92ad873e54917b8c419c/src/bounce_solution/bounce_solution.cpp#L757) and the condition in [lines 784-788](https://github.com/phbasler/BSMPT/blob/0aea6b73e6dd2165996b92ad873e54917b8c419c/src/bounce_solution/bounce_solution.cpp#L784) to exit the loop are the same after the changes of the PR. It is thus probably not necessary to keep both?

Maybe, if I may make a suggestion, a safer way to handle this loop could be to do the following:
```
    while (not(std::abs(T_up / T_down - 1) <
                   RelativeTemperatureInCalcTempAtFalseVacFraction or
               almost_the_same(int_at_false_vac_frac,
                               IatT,
                               RelativeErrorInCalcTempAtFalseVacFraction)))
    {
      T_middle = (T_up + T_down) / 2.;
      this->SetStoredTemp(T_middle); // update temp storage for inner integral
      IatT = prefac * Nintegrate_Outer(*this).result;

      Logger::Write(LoggingLevel::BounceDetailed,
                    "Pf ( T = " + std::to_string(T_middle) +
                        " ) = " + std::to_string(std::exp(-IatT)));

      if (IatT < int_at_false_vac_frac)
      {
        T_up = T_middle;
      }
      else
      {
        T_down = T_middle;
      }

      res_Temp = T_middle;
    }
```
i.e. change the `and` in the `while` condition to an `or` (so that the loop exits when at least one of the conditions is fulfilled, and not wait for both), and remove the `if` condition at the end of the loop, just keep the `res_Temp = T_middle;` line. This should not change anything regarding the output, and will make sure that, as soon as the `while` loop is entered, a valid temperature will always be output, and the `while` loop will always exit after a certain number of iterations (when `T_up` and `T_down` are close enough). Unless it is not desired that upon entering the `while` loop, the function `CalcTempAtFalseVacFraction` always returns a valid temperature?

vollous committed:
Fix 224-problem-with-conan-due-to-missing-call-of-setup_cmaes-in-buildpy

lisabiermann committed:
Fix user_input description

lisabiermann committed:
Merge branch 'develop' into Fix-user-input-description

vollous committed:
Merge branch 'master' into develop

lisabiermann committed:
Merge branch 'develop' into fix_Temp_from_VacFraction

phbasler committed:
Merge branch 'master' into fix_Temp_from_VacFraction

Commit:
Automatic version bump (patch). 3.1.11 -> 3.1.12

vollous commented:
Superseded by https://github.com/BSMPT/BSMPT/pull/295
